### PR TITLE
adding qtum connector

### PR DESCRIPTION
**Pull request by Rachit-gaur**

### CODE REVIEW

File: qtum-connector/module/package.json [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    "access": "public"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/tatumio/tatum-blockchain-connector/qtum-connector"
+  },
+  "bugs": "https://github.com/tatumio/tatum-blockchain-connector/",
+  "peerDependencies": {
+    "@nestjs/common": "^7.6.1",
+    "@tatumio/tatum": "^1.14.5",
```
Comment (samuelsramko) [inline qtum-connector/module/package.json:38]: increase tatumjs lib peer dependency version to the latest one

File: qtum-connector/module/src/QtumController.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import { QtumService } from './QtumService';
+import { QtumError } from './QtumError';
+import { QtumEndpointGuard } from './QtumEndpointGuard';
+import { RawTx, contractCall, QtumIRawTransactionInfo, QtumIRawTransactions } from './dto/qtumTx';
+
+@UseGuards(QtumEndpointGuard)
+export abstract class QtumController {
+    protected constructor(protected readonly service: QtumService) {
+    }
+    @Post('v3/qtum/contractCall')
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:11]: please unify paths according to other blockchain integrations

File: qtum-connector/module/src/QtumController.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    async contractCall(@Body() body: contractCall) {
+        try {
+            return await this.service.contractCall(body.testnet, body.address, body.encodedData);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Post('v3/qtum/sendRawTx')
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:19]: please unify paths according to other blockchain integrations

File: qtum-connector/module/src/QtumController.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import { RawTx, contractCall, QtumIRawTransactionInfo, QtumIRawTransactions } from './dto/qtumTx';
+
+@UseGuards(QtumEndpointGuard)
+export abstract class QtumController {
+    protected constructor(protected readonly service: QtumService) {
+    }
+    @Post('v3/qtum/contractCall')
+    async contractCall(@Body() body: contractCall) {
+        try {
+            return await this.service.contractCall(body.testnet, body.address, body.encodedData);
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:14]: testnet is never a parameter inside the request. It is obtained from the API key

File: qtum-connector/module/src/QtumController.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import { QtumError } from './QtumError';
+import { QtumEndpointGuard } from './QtumEndpointGuard';
+import { RawTx, contractCall, QtumIRawTransactionInfo, QtumIRawTransactions } from './dto/qtumTx';
+
+@UseGuards(QtumEndpointGuard)
+export abstract class QtumController {
+    protected constructor(protected readonly service: QtumService) {
+    }
+    @Post('v3/qtum/contractCall')
+    async contractCall(@Body() body: contractCall) {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:12]: contractCall should start with capital C

File: qtum-connector/module/src/QtumController.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    @Post('v3/qtum/sendRawTx')
+    async sendRawTx(@Body() body: RawTx) {
+        try {
+            return await this.service.sendRawTx(body.testnet, body.rawtx);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+
+    @Get('/v3/qtum/:chain/utxos/:address')
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:28]: chain != testnet. should be ommited

File: qtum-connector/module/src/QtumController.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    async getQtumUTXOs(@Param() path: { chain: string, address: string }) {
+        try {
+            return await this.service.getQtumUTXOs(path.chain == 'testnet' ? true : false, path.address);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('/v3/qtum/:chain/:address/info')
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:36]: unify path with other account detail endpoints

File: qtum-connector/module/src/QtumController.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    @Get('/v3/qtum/:chain/transaction/:id')
+    async getQtumTransaction(@Param() path: { id: string, chain: string }): Promise<QtumIRawTransactionInfo> {
+        try {
+            return await this.service.getQtumTransaction(path.chain == 'testnet' ? true : false, path.id);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('/v3/qtum/:chain/transactions/:id/:pageNum')
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:52]: use common Pagination.ts with pageSize and offset

Comment (samuelsramko) [reply qtum-connector/module/src/QtumController.ts:52]: what is id parameter here?

File: qtum-connector/module/src/QtumEndpointGuard.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import {CanActivate, ExecutionContext} from '@nestjs/common';
+import {Request} from 'express';
+import {QtumError} from './QtumError';
+import {QTUM_HEADER_ENDPOINT} from './index';
+
+export class QtumEndpointGuard implements CanActivate {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumEndpointGuard.ts:6]: this guard does not have to be here. It's not like Quorum or Hyperledger, where everyone could have its own instance of private nodes. Please delete this.

File: qtum-connector/module/src/QtumService.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+    public async contractCall(
+        testnet: boolean,
+        address: string,
+        encodedData: string,
+    ): Promise<QtumIContractCall> {
+        try {
+            const baseURL = testnet ? INSIGHT_BASEURLS['QTUM_TESTNET'] : INSIGHT_BASEURLS['QTUM_MAINNET']
+            const res = await axios.get(baseURL +
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:53]: we should do this contrctCall more easier for devs. They don't know how to create encoded data etc. Please take a look on how we do it in ETH or BSC - InvokeSmartContractMethod.

File: qtum-connector/module/src/dto/qtumTx.ts [hunk 11]
```diff
@@ -0,0 +1,155 @@
+import { Currency } from '@tatumio/tatum';
+export class QtumIRawTransactionInfo {
```
Comment (samuelsramko) [inline qtum-connector/module/src/dto/qtumTx.ts:2]: these classes / interfaces should be split into separate files

File: qtum-connector/module/src/index.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+export * from './QtumService';
+export * from './QtumController';
+export * from './QtumError';
+export const QTUM_HEADER_ENDPOINT = 'x-qtum-endpoint';
```
Comment (samuelsramko) [inline qtum-connector/module/src/index.ts:4]: not necessary

File: qtum-connector/module/src/index.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+export const INSIGHT_BASEURLS: { [key: string]: string } = {
```
Comment (samuelsramko) [inline qtum-connector/module/src/index.ts:5]: renme to sthing more clear, like QTUM_URLS.

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: qtum-connector/module/src/QtumController.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+    @Get('/v3/qtum/utxo/:address')
+    async getQtumUTXOs(@Param() address: string ) {
+        try {
+            return await this.service.getQtumUTXOs(address);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('/v3/qtum/:address/balance')
+    async getInfo(@Param() address: string) {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:68]: rename oath to v3/qtum/address/:address/balance

File: qtum-connector/module/src/QtumController.ts [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    @Get('/v3/qtum/transaction/:id')
+    async getQtumTransaction(@Param() id: string): Promise<QtumIRawTransactionInfo> {
+        try {
+            return await this.service.getQtumTransaction(id);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('/v3/qtum/transactions/address/:address/:pageNum')
+    async getQtumTransactions(@Param() path: { address: string, pageNum: number }): Promise<QtumIRawTransactions> {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:84]: pagination should be done in a unified way. We have query parameter pageSize and offset in Pagination.ts file. Please use this aproach.

File: qtum-connector/module/src/QtumService.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+            const data = () => ({ "jsonrpc": "1.0", "method": "getbestblockhash", "params": [] })
+            const config = { headers: { 'content-type': 'text/plain;' } }
+            const res = await axios.post(baseurl, data(), config)
+            return res['result']
+        } catch (e) {
+            this.logger.error(e)
+            throw new QtumError(`Error occurred. ${e}`, 'qtum.error');
+        }
+    }
+    public async getBlock(hash: string): Promise<QtumBlock> {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:61]: this method should cover get block by hash and get block by height. Is this supported now?

File: qtum-connector/module/src/QtumService.ts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        } catch (e) {
+            this.logger.error(e)
+            throw new QtumError(`Error occurred. ${e}`, 'qtum.error');
+        }
+    }
+
+    public async broadcast(rawtx: string): Promise<QtumISendRawTxResult> {
+        try {
+            const network = await this.isTestnet()
+            const baseURL = network ? QTUM_URLS['QTUM_TESTNET'] : QTUM_URLS['QTUM_MAINNET']
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:77]: url should be obtained from getFirstNodeUrl()

File: qtum-connector/module/src/QtumService.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+            return res.data
+        } catch (e) {
+            this.logger.error(e)
+            throw new QtumError(`Error occurred. ${e}`, 'qtum.error');
+        }
+    }
+    public async estimateFee(nblocks: number = 6): Promise<any> {
+        try {
+            const network = await this.isTestnet()
+            const baseURL = network ? QTUM_URLS['QTUM_TESTNET'] : QTUM_URLS['QTUM_MAINNET']
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:91]: url should be obtained from getFirstNodeUrl()

File: qtum-connector/module/src/QtumService.ts [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+            throw new QtumError(`Error occurred. ${e}`, 'qtum.error');
+        }
+
+    }
+    public async getQtumTransaction(
+        id: string,
+    ): Promise<QtumIRawTransactionInfo> {
+        try {
+            const network = await this.isTestnet()
+            const baseURL = network ? QTUM_URLS['QTUM_TESTNET'] : QTUM_URLS['QTUM_MAINNET']
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:123]: url should be obtained from getFirstNodeUrl()

File: qtum-connector/module/src/QtumService.ts [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+    }
+    public async getQtumTransactions(
+        address: string,
+        pageNum: number = 0,
+    ): Promise<QtumIRawTransactions> {
+        try {
+            const network = await this.isTestnet()
+            const baseURL = network ? QTUM_URLS['QTUM_TESTNET'] : QTUM_URLS['QTUM_MAINNET']
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumService.ts:138]: url should be obtained from getFirstNodeUrl()

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: qtum-connector/module/src/QtumService.ts (hunk 16 above)
Comment (Rachit-gaur) [reply qtum-connector/module/src/QtumService.ts:61]: get block by hash is covered, and getCurrentBlock will give the hash of latest block.
for only these 2 functions I have used the node that we have hosted for qtum

File: qtum-connector/module/src/QtumService.ts (hunk 17 above)
Comment (Rachit-gaur) [reply qtum-connector/module/src/QtumService.ts:77]: actually this is not from the node but we are consuming the insight api to fetch these details, getFirstNodeUrl() will be used to fetch nodes, do I enter all these urls there itself?

Comment (Rachit-gaur) [reply qtum-connector/module/src/QtumService.ts:77]: shall I make all the calls from the node itself and not this api?

File: qtum-connector/module/src/QtumController.ts [hunk 21]
```diff
@@ -32,23 +32,23 @@ export abstract class QtumController {
         }
     }
     @Get('v3/qtum/block/:hash')
-    async getBlock(@Param() hash:string) {
+    async getBlock(@Param() hash: string) {
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:35]: you are not validating Params in any request. Take a look how it is done on another connectors. Query and Path params must be validated the same way as request bodies

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: qtum-connector/module/src/QtumController.ts [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+import { PathAddress } from './dto/PathAddress';
+import { PathId } from './dto/PathId';
+import { Pagination } from './dto/Pagination';
+export abstract class QtumController {
+    protected constructor(protected readonly service: QtumService) {
+    }
+    @Post('v3/qtum/wallet/priv')
+    async generatePrivateKey(@Body() body: GeneratePrivateKey) {
+        try {
+            return await generatePrivateKeyFromMnemonic(Currency.QTUM, true, body.mnemonic, body.index);
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:20]: you have true here - shoud be await this.isTestnet()

File: qtum-connector/module/src/QtumController.ts [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('v3/qtum/address/:xpub/:i')
+    async generateAddress(@Param() { xpub, i }: PathXpubI) {
+        try {
+            return await generateAddressFromXPub(Currency.QTUM, true, xpub, Number(i));
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:28]: you have true here - shoud be await this.isTestnet()

File: qtum-connector/module/src/QtumController.ts [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+        try {
+            return await this.service.getBlock(hash);
+        } catch (e) {
+            throw new QtumError(`Unexpected error occurred. Reason: ${e.message || e.response?.data || e}`, 'qtum.error');
+        }
+    }
+    @Get('v3/qtum/address/:key')
+    async generateAddressPrivatekey(@Param() { key }: PathKey) {
+        try {
+            return await generateAddressFromPrivatekey(Currency.QTUM, true, key);
```
Comment (samuelsramko) [inline qtum-connector/module/src/QtumController.ts:52]: you have true here - shoud be await this.isTestnet()

**Review by samuelsramko — CHANGES_REQUESTED** [review]
I dont see openapi.yaml file anywhere, which describes the API

File: qtum-connector/module/src/openapi.yaml [hunk 25]
```diff
@@ -0,0 +1,879 @@
+openapi: 3.0.0
+servers:
+  - url: 'https://api-eu1.tatum.io'
```
Comment (samuelsramko) [inline qtum-connector/module/src/openapi.yaml:3]: you have a lot of errors here. Try to paste the content to https://editor.swagger.io/ and you will see.

**Review by samuelsramko — CHANGES_REQUESTED** [review]

**Review by samuelsramko — APPROVED** [review]

**Pull request merged**